### PR TITLE
Updated anki-sync-server to work with the latest version of Anki

### PR DESCRIPTION
**Pull request by kalehmann**
This PR tackles the following issues: #13, #14, #16 and #19.

* The embedded Anki submodule was removed, anki-sync-server now supports the current version of Anki installed by `pip install anki`.
* To support newer versions of Anki, a Nginx-Proxy to "unchunk" requests is required before the anki-sync-server
* An example addon for Anki versions >= 2.1.28 is provided in the Readme
* The web media test was removed, because it depended on the MediaSyncer class, which no longer exists

I tested these changes with the latest master of AnkiDroid and Anki 2.1.28. I have found no issues while syncing between these two.

## Required steps before merging

- [x] Patching all tests to work with the new code
- [x] Write Anki dependency to lock file

### CODE REVIEW

**Comment by AlexBocken** [comment]
> - To support newer versions of Anki, a Nginx-Proxy to "unchunk" requests is required before the anki-sync-server

If I try to run the example unchunker as given in the Readme using nginx, I'm not able to run the ankisyncd server at the same time via `python -m ankisyncd` as this Error: `OSError: [Errno 98] Address already in use` get's raised.

Which seems reasonable to me as an error. Of course this could be a result of my incompetence with nginx, I'm rather new to it. Does `ankisyncd.conf` need to modified to use a different port? But then the `proxy_pass` setting in the given example nginx setup does not make sense to me anymore.

Would it be possible to update the documentation to further clarify this area for an easier implementation? (I'm already quite grateful for the example nginx config at all). My guess is that other's who will try to to implement this workaround will have similar issues if they're not too familiar with nginx.

I apologize if this is the wrong place for a comment like this, but I didn't feel like it's appropriate to raise an issue on this project and the issues feature seems to be disabled on your personal fork.

**Comment by kalehmann** [comment]
HI @AlexBocken, thank you for your input.

> I apologize if this is the wrong place for a comment like this, but I didn't feel like it's appropriate to raise an issue on this project and the issues feature seems to be disabled on your personal fork.

Your comment is suited perfectly fine here as it describes a problem introduced by this pull request 😄 

During development I have used docker both services (ankisyncd and Nginx) ran on different containers. Therefore they could use the same port.  As you have already found out, this is not possible when both services run on your local machine.

I have updated the ready and further clarified how Nginx should be configured when both services run on the same machine.

**Comment by AlexBocken** [comment]
Thanks for the clarifications!

File: .gitignore [hunk 1]
```diff
@@ -95,6 +95,10 @@ share/python-wheels/
 *.egg
 MANIFEST
 
+# Emacs temporary files
```
Comment (VikashKothary) [inline .gitignore:98]: 👍

File: README.md [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ankisyncd
-=========
+11;rgb:2323/2727/2929=========
```
Comment (VikashKothary) [inline README.md:2]: Is this a typo?

File: README.md [hunk 3]
```diff
@@ -39,22 +39,8 @@ It supports Python 3 and Anki 2.1.
...
-
-        $ git submodule update --init
-        $ cd anki-bundled
-        $ pip install -r requirements.txt
-
-   Keep in mind `pyaudio`, a dependency of Anki, requires development headers for
-   Python 3 and PortAudio to be present before running `pip`. If you can't or
-   don't want to install these, you can try [patching Anki](#running-ankisyncd-without-pyaudio).
-
+0. Install the current version of Anki.
```
Comment (VikashKothary) [inline README.md:42]: This is more on me but https://github.com/ankicommunity/anki-sync-server/issues/23. Step 0 and 1 should be combined to `pip install -r src/requirements.txt`. Not a blocker for this PR but just a comment.

File: README.md [hunk 4]
```diff
@@ -65,21 +51,38 @@ Installing
 
         $ ./ankisyncctl.py adduser <username>
 
-4. Run ankisyncd:
+4. Setup a proxy to unchunk the requests.
```
Comment (VikashKothary) [inline README.md:54]: Definitely very interesting.

File: README.md [hunk 5]
```diff
@@ -65,21 +51,38 @@ Installing
...
+    For example, if you use Nginx  on the same machine as ankisyncd, you first
+    have to change the port in `ankisyncd.conf` to something other than `27701`.
+    Then configure Nginx to listen on port `27701` and forward the unchunked
+    requests to ankisyncd.
+
+    An example configuration with ankisyncd running on the same machine as Nginx
+    and listening on port `27702` may look like:
 
-<span id="readme-fn-01"></span>
-1. 2.1.9 is not supported due to [commit `95ccbfdd3679`][] introducing the
```
Comment (VikashKothary) [inline README.md:75]: Does this mean version 2.1.9 is not supported?

File: src/ankisyncd/sync_app.py [hunk 6]
```diff
@@ -35,24 +35,25 @@
...
-import anki.sync
 import anki.utils
-from anki.consts import SYNC_VER, SYNC_ZIP_SIZE, SYNC_ZIP_COUNT
 from anki.consts import REM_CARD, REM_NOTE
 
 from ankisyncd.users import get_user_manager
 from ankisyncd.sessions import get_session_manager
 from ankisyncd.full_sync import get_full_sync_manager
 
+from .sync import Syncer, SYNC_VER, SYNC_ZIP_SIZE, SYNC_ZIP_COUNT
```
Comment (VikashKothary) [inline src/ankisyncd/sync_app.py:45]: Can you use an absolute import: `from ankisyncd import ...`

File: README.md [hunk 7]
```diff
@@ -125,8 +140,6 @@ Unless you have set up a reverse proxy to handle encrypted connections, use
 whatever you have specified in `ankisyncd.conf` (or, if using a reverse proxy,
 whatever port you configured to accept the front-end connection).
 
-**Do not use trailing slashes.**
```
Comment (VikashKothary) [inline README.md:128]: Just confirming you've fixed this?

**Review by VikashKothary — COMMENTED** [review]
Mate this is great work. It's all looks good to me. Thanks for all your hard work. A few quick changes and this PR should be ready to be merged.

File: README.md (hunk 7 above)
Comment (kalehmann) [reply README.md:128]: Yes, I have fixed this with ankidroid/Anki-Android#6437

File: src/ankisyncd/sync_app.py (hunk 6 above)
Comment (AntonOfTheWoods) [reply src/ankisyncd/sync_app.py:45]: Agreed, I think both are fine but like 'tabs or spaces?', the correct answer is 'whatever is there when you started, unless you are prepared to convince all current contributors and migrate everything, personally supporting any and all issues that arise' :-).

Comment (kalehmann) [reply src/ankisyncd/sync_app.py]: Yes, fixed it.

File: README.md (hunk 2 above)
Comment (kalehmann) [reply README.md]: fixed

Comment (kalehmann) [reply README.md]: fixed

**Review by AntonOfTheWoods — CHANGES_REQUESTED** [review]
There is some code that looks like it is no longer needed. AFAICT, for example, there is no longer any particular use for the `runHook`s. The `sync` method is also, unless I'm mistaken, client-side code. As we now have to maintain this, it is very important not to have dead code that will just confuse future contributors. Can you give the file a clean?

File: README.md (hunk 3 above)
Comment (kalehmann) [reply README.md]: Okay, I have updated the Readme

File: README.md (hunk 5 above)
Comment (kalehmann) [reply README.md:75]: I believe this restriction only applied for the server side version of Anki. With the changes introduced by this pull request we require the user to have the latest version of Anki available thorugh pip installed. 

Therefore this section of the Readme serves no purpose anymore.

**Review by AntonOfTheWoods — COMMENTED** [review]
Are you deleting code you added in https://github.com/ankicommunity/anki-sync-server/pull/20/commits/59603b832db31ddd5959e584a4450812e412a1aa ? If so, can you combine the commits instead? If not, I'm lost, could you explain? :-)

**Comment by AntonOfTheWoods** [comment]
Thanks again for the great work @kalehmann ! Apart from a couple of minor issues it looks great. I know it's sometimes hard and a little annoying but maybe next time maybe try and have several smaller PRs to avoid changing the same parts of the code in several commits of the PR.

File: src/ankisyncd/full_sync.py [hunk 8]
```diff
@@ -1,45 +1,81 @@
...
-import anki.db
+from anki.db import DB
+from anki.collection import Collection
 
-class FullSyncManager:
-    def upload(self, col, data, session):
+logger = logging.getLogger("ankisyncd.media")
+logger.setLevel(1)
+
+class FullSyncManager(object):
```
Comment (AntonOfTheWoods) [inline src/ankisyncd/full_sync.py:16]: Do you have a reason for preferring having an explicit inheritance of (object)? In python3 it makes no difference and it wasn't used before, so I'm not sure I see the value. In any case, I don't think we should mix (having and not having) and I personally don't think it serves a great purpose.

Comment (kalehmann) [reply src/ankisyncd/full_sync.py:16]: Just a relict from the good old Python 2 days, when inheriting from `object` did indeed make a difference.

**Comment by kalehmann** [comment]
> Are you deleting code you added in [59603b8](https://github.com/ankicommunity/anki-sync-server/commit/59603b832db31ddd5959e584a4450812e412a1aa) ? If so, can you combine the commits instead? 

Yes, it seems like Anki desktop only  accepts the downgraded database, therefore I reverted these changes. Anyways, I squashed these two commits

**Comment by kalehmann** [comment]
> I know it's sometimes hard and a little annoying but maybe next time maybe try and have several smaller PRs to avoid changing the same parts of the code in several commits of the PR.

Yeah I know 🤔 The problem is just, that when I started working on this PR I didn't even know how big it would get. I believe splitting it into smaller changes requires a profound evaluation of all necessary changes beforehand - which I was unable to do because I knew nothing about this project.

**Comment by AntonOfTheWoods** [comment]
> Yes, it seems like Anki desktop only accepts the downgraded database, therefore I reverted these changes. Anyways, I squashed these two commits

That's interesting, I didn't follow that up for `djankiserv`, it "Just Worked", and I didn't really know why :-). I definitely think that they will fully migrate to the new structure soon though, so that is definitely a TODO to test properly.

**Comment by AntonOfTheWoods** [comment]
> I have removed any code related to these two points. But I still believe, that there is more unused code in this project. However I think that finding and removing these lines would better suite another PR.

Oh, how many times I would have liked to have got away with that at my last company... :-). The way I am managing this for importing upstream code to `djankiserv` is to bring in a module then write tests to get the code to do useful stuff. Then run `coverage` and anything that hasn't been run gets the chop. It is pretty violent and misses cases but we can always go back and get code if it really was necessary. Let's definitely plan for something like this!

**Review by AntonOfTheWoods — APPROVED** [review]
LGTM. There is a lot of cleanup to be done but there was before so let's get this merged and move forward from there.

**Pull request merged**